### PR TITLE
feat(probabilistic_occupancy_grid_map): add projective raytracing option from scan_origin

### PR DESCRIPTION
**Pull request by soblin**
## Description

Modified/Fixed pointcloud based ocucpancy grid_map to consider the projected length of obstacle point clouds from scan_origin to estimate long range unknown region better.

## Related links

launcher PR: https://github.com/autowarefoundation/autoware_launch/pull/434

https://tier4.atlassian.net/browse/RT1-2838

## Tests performed

- Psim works as before
- logging_simulator works as before

image

## Interface changes

Not applicable.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [X] I've confirmed the [contribution guidelines].
- [X] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [X] The PR follows the [pull request guidelines].
- [X] The PR has been properly tested.
- [X] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [X] There are no open discussions or they are tracked via tickets.
- [X] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml [hunk 1]
```diff
@@ -18,4 +18,10 @@
     # center of the grid map
     gridmap_origin_frame: "base_link"
     # ray-tracing center: main sensor frame is preferable like: "velodyne_top"
-    scan_origin_frame: "base_link"
+    scan_origin_frame: "velodyne_top"
+
+    use_projection: true
+    projection_dz_threshold: 0.01 # [m] for avoiding null division
+    obstacle_separation_threshold: 1.0 # [m] fill the interval between obstacls with unknown for this length
```
Comment (yukkysaito) [inline perception/probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml:25]:
```suggestion
    obstacle_separation_threshold: 1.0 # [m] fill the interval between obstacles with unknown for this length
```

File: perception/probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml [hunk 2]
```diff
@@ -18,4 +18,10 @@
     # center of the grid map
     gridmap_origin_frame: "base_link"
     # ray-tracing center: main sensor frame is preferable like: "velodyne_top"
-    scan_origin_frame: "base_link"
+    scan_origin_frame: "velodyne_top"
```
Comment (yukkysaito) [inline perception/probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml:21]: The frame velodyne top is not defined as an official AWF tf, so could you please make it a base_link?

Comment (soblin) [reply perception/probabilistic_occupancy_grid_map/config/pointcloud_based_occupancy_grid_map.param.yaml:21]: I see, then I will revert this flag to false and scan_origin_frame to "base_link", and only use "velodyne_top" link name for internal use only.

File: perception/probabilistic_occupancy_grid_map/include/pointcloud_based_occupancy_grid_map/occupancy_grid_map.hpp [hunk 3]
```diff
@@ -70,10 +71,15 @@ class OccupancyGridMap : public nav2_costmap_2d::Costmap2D
...
-  void updateWithPointCloud(
+  void updateWithPointCloudSimple(
     const PointCloud2 & raw_pointcloud, const PointCloud2 & obstacle_pointcloud,
     const Pose & robot_pose, const Pose & scan_origin);
 
+  void updateWithPointCloud3D(
+    const PointCloud2 & raw_pointcloud, const PointCloud2 & obstacle_pointcloud,
+    const Pose & robot_pose, const Pose & scan_origin, const double projection_dz_threshold,
+    const double obstacle_separation_threshold, grid_map::GridMap * debug);
+
```
Comment (yukkysaito) [inline perception/probabilistic_occupancy_grid_map/include/pointcloud_based_occupancy_grid_map/occupancy_grid_map.hpp:82]: Can you define interface classes and have them inherit? (like strategy pattern)

File: perception/probabilistic_occupancy_grid_map/launch/pointcloud_based_occupancy_grid_map.launch.py [hunk 4]
```diff
@@ -73,13 +81,25 @@ def launch_setup(context, *args, **kwargs):
...
     )
 
+    grid_map_visualization_node = Node(
+        package="grid_map_visualization",
+        executable="grid_map_visualization",
+        name="grid_map_visualization",
+        output="screen",
+        parameters=[grid_map_params],
+    )
+
```
Comment (yukkysaito) [inline perception/probabilistic_occupancy_grid_map/launch/pointcloud_based_occupancy_grid_map.launch.py:91]: I would like to run only debug mode.

File: perception/probabilistic_occupancy_grid_map/include/pointcloud_based_occupancy_grid_map/occupancy_grid_map.hpp (hunk 3 above)
Comment (soblin) [reply perception/probabilistic_occupancy_grid_map/include/pointcloud_based_occupancy_grid_map/occupancy_grid_map.hpp:82]: @yukkysaito Originally I created a new class `projective_pointcloud_based_occupancy_grid_map` and specified that from the launcher. Do you mean this mechanism ?

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**